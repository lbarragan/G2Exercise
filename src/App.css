.App {
  text-align: center;
}

.App h1 {
  padding-top: 15px;
  color: #000000;
  -webkit-text-fill-color: #ffffff;
  -webkit-text-stroke-width: 1px;
  -webkit-text-stroke-color: navy;
}

.App-Content{
  width: 600px;
  height: 300px;
  background: #ffffff;
  margin: 40px auto;
}

.App-Content-input{
  border: 0;
  border-bottom: 2px solid navy;
  width: 200px;
  font-size: 14px;
  text-align: left;
  padding: 5px;
  background: transparent;
  color: navy;
}

.App-Content-input-invalid{
  border: 0;
  border-bottom: 2px solid #B80000;
  width: 200px;
  font-size: 14px;
  text-align: left;
  padding: 5px;
  background: transparent;
  color: #E80000;
}

.App-Content-description{
  color: #A9A9A9;
  font-size: 12px;
  text-align: left;
  padding-left: 140px;
}

.App-Content-errorMsg{
  color: #E80000;
  font-size: 12px;
  position: absolute;
  top: 115px;
  left: 140px;
}

.App-Content input[type="button"]{
  border-radius: 10px;
  padding: 5px 12px;
  font-size: 14px;
  background-color: navy;
  color:white;
  border-color: navy;
}

.App-Content input[type="button"]:disabled{
  border-radius: 10px;
  padding: 5px 12px;
  font-size: 14px;
  background-color: #D3D3D3;
  color:white;
  border-color: #A9A9A9;
}

.App-Content-Weather-Box{
  position: absolute;
  overflow-x: auto;
}

.App-Content-Weather-Box ul{
  list-style: none;
}

.App-Content-Weather-Box li{
  display: inline-block;
  vertical-align: top;
  padding: 10px;
}

.App-Content-Weather-Day{
  font-weight: bolder;
  font-size: 14px;
  color: navy;
}

.App-Content-Weather-Temp{
  font-weight: bold;
  font-size: 10px;
  color: navy;
}

/********************
* Shadow Effect
*********************/
.shadowEffect{
  position: relative;
  box-shadow: 0 1px 4px rgba(0, 0, 0, 0.3), 0 0 40px rgba(0, 0, 0, 0.1) inset;
}

.shadowEffect:before, .shadowEffect::after{
  content: "";
  position: absolute;
  z-index: -1;
  box-shadow: 0 0 20px rgba(0, 0, 0, 0.8);
  top: 10px;
  bottom: 10px;
  left: 0;
  right: 0;
  border-radius: 100px/10px;
}

.shadowEffect::after{
  right: 10px;
  left: auto;
  transform: skew(8deg) rotate(3deg);
}