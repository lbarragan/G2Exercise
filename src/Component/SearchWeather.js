import React from 'react';

export default class SearchWeather extends React.Component{
    constructor(props){
        super(props);
        this.state = {
            city: '',
            isValid: props.isValid,
            errorMsg: props.errorMsg
        }

        this.handleCityChange = this.handleCityChange.bind(this);
        this.handleSearch = this.handleSearch.bind(this);
    }

    setError(message){
        this.setState({
            errorMsg: message.charAt(0).toUpperCase() + message.slice(1).toLowerCase()
        });
    }

    handleCityChange(event){
        let regex = /^([a-zA-Z ])+\s*(,)\s*([a-zA-Z]{2})$/;
        let validationMsg = '';
        let isValid = true;
        if(!regex.test(event.target.value)){
            validationMsg = 'Invalid City format';
            isValid = false;
        }
        this.setState({ city: event.target.value, isValid:isValid, errorMsg: validationMsg });
    }

    handleSearch(event){
        if(this.state.city.trim() === ""){
            this.setState({
                errorMsg: 'City cannot be empty',
                isValid: false
            })
        }else{
            this.props.ValidateCity(event, this.state.city);
        }
    }

    render(){
        return (
            <div>
                <form>
                    <p>
                       <input type="text" name="city" value={this.state.name} onChange={this.handleCityChange} className={this.state.isValid ? "App-Content-input" : "App-Content-input-invalid"} placeholder="City name, Country code" />&nbsp;&nbsp;&nbsp;&nbsp;
                       <input type="button" value="Take a look!" onClick={this.handleSearch} disabled={!this.state.isValid} />
                    </p>
                    <span className="App-Content-errorMsg">{this.state.errorMsg}</span>
                </form>
            </div>
        )
    }
}