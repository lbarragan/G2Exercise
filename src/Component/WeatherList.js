import React from 'react';

export default class WeatherList extends React.Component{
    render(){
        return(
            <div className="App-Content-Weather-Box">
                {
                    this.props.Data.length === 0 
                        ? <div className="App-Content-description">
                            *Enter City name and country code divided by comma<br />
                            *Use ISO 3166 country codes<br />
                            *E.g. <b>New York, US</b>
                          </div>
                        : <ul>
                            {
                                this.props.Data.map(({date, day, min, max, weather, image}) => 
                                    <li key={date}>
                                        <div>
                                            <div className="App-Content-Weather-Day">{day}</div>
                                            <div><img src={image} alt={weather} /></div>
                                            <div className="App-Content-Weather-Temp">
                                                <span>{min}&#176;F&nbsp;&nbsp;&nbsp;&nbsp;</span>
                                                <span>{max}&#176;F</span>
                                            </div>
                                        </div>
                                    </li>
                                )
                            }
                          </ul>
                }
            </div>
        );
    }
}