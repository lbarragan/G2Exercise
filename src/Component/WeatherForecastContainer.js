import React from 'react';
import SearchWeather from './SearchWeather';
import WeatherList from './WeatherList';

export class WeatherForecastContainer extends React.Component{
    constructor(props){
        super(props);
        this.state = {
            data: [],
            isValid: true,
            errorMsg: ' '
        }

        this.ValidateCity = this.ValidateCity.bind(this);
        this.searchWeather = React.createRef();
    }

    ValidateCity(event, city){
        event.preventDefault();
        fetch('http://api.openweathermap.org/data/2.5/forecast?q=' + city + '&APPID=afbca948c1936057aede841310422cd9&units=imperial', {method: 'GET'})
        .then((response) => {
            if(response.headers.get("content-type").indexOf("application/json") !== -1){
                return response.json();
            }
        }).then((responseJson) => {
            if(responseJson.cod === "200"){
                let weathers = [];
                let days = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
                for(let i=0; i < responseJson.list.length; i++){
                    let currentTime = new Date();
                    while(currentTime.getHours() % 3 !== 0){
                        currentTime.setHours(currentTime.getHours() - 1);
                    }
                    let d = new Date(responseJson.list[i].dt_txt);
                    if(currentTime.getHours() === d.getHours()){
                        let weather = {
                            "date": responseJson.list[i].dt,
                            "day": days[d.getDay()],
                            "min": responseJson.list[i].main.temp_min,
                            "max": responseJson.list[i].main.temp_max,
                            "weather": responseJson.list[i].weather[0].main,
                            "image": "http://openweathermap.org/img/w/" + responseJson.list[i].weather[0].icon + ".png"
                        }
                        
                        weathers.push(weather);
                    }
                }
                this.setState({
                    data: weathers,
                    isValid: true,
                    errorMsg: ''
                }); 
            }else{
                this.searchWeather.current.setError(responseJson.message);
            }
        })
        .catch(error => console.log(error));
    }

    render(){
        return(
            <div>
                <SearchWeather ref={this.searchWeather} ValidateCity={this.ValidateCity} isValid={this.state.isValid} errorMsg={this.state.errorMsg} />
                <WeatherList Data={this.state.data} />
            </div>
        )
    }
}