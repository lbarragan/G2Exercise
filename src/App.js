import React from 'react';
import './App.css';
import { WeatherForecastContainer } from './Component/WeatherForecastContainer';

function App() {
  return (
    <div className="App">
      <div className="App-Content shadowEffect">
        <h1>G2 Weather Forecast</h1>
        <WeatherForecastContainer />
      </div>
    </div>
  );
}

export default App;
