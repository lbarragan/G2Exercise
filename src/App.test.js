import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import Adapter from 'enzyme-adapter-react-16'
import { shallow, configure, mount } from "enzyme";
import { WeatherForecastContainer } from './Component/WeatherForecastContainer';
import SearchWeather from './Component/SearchWeather';
import WeatherList from './Component/WeatherList';

configure({ adapter: new Adapter()});

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});

describe("G2 Weather Forecast Exercise", ()=> {
  let props;
  let days = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
  let d = new Date(); 
    
  beforeEach(() => {
    props = {
      data: [{
        "date": d.getUTCDate(),
        "day": days[d.getDay()],
        "min": "60",
        "max": "70",
        "weather": "Clear",
        "image": "http://openweathermap.org/img/w/01d.png"
      }],
      isValid: true,
      errorMsg: ''
    };
  });

  it("should render correctly with no props", ()=>{
    const wrapper = shallow(<WeatherForecastContainer />);
    expect(wrapper).toMatchSnapshot();
  });

  it("Always renders a 'SearchWeather'", () =>{
    const wrapper = mount(<WeatherForecastContainer {...props} />);
    expect(wrapper.children().first(SearchWeather).length).toBe(1);
  });

  it("'SearchWeather' receive one prop", () =>{
    const wrapper = mount(<WeatherForecastContainer {...props} />);
    const swWrapper = wrapper.children().first(SearchWeather);
    expect(Object.keys(swWrapper.props()).length).toBe(1);
  });

  it("Always renders a 'WeatherList'", () =>{
    const wrapper = mount(<WeatherForecastContainer {...props}/>);
    expect(wrapper.children().contains(WeatherList)).toBe(true);
  });

  it("'WeatherList' receive one prop", () =>{
    const wrapper = mount(<WeatherForecastContainer {...props} />);
    const listWrapper = wrapper.children().first(WeatherList);
    expect(Object.keys(listWrapper.props()).length).toBe(1);
  });

});