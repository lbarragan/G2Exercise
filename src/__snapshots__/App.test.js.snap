// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`G2 Weather Forecast Exercise should render correctly with no props 1`] = `ShallowWrapper {}`;
